### PR TITLE
Generate static callbacks

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

**Comment by Therzok** [comment]
Generated changes diff https://gist.github.com/Therzok/220d717136437919b9978466db46b678

**Comment by Therzok** [comment]
https://gist.github.com/Therzok/74bb08f541d7fa1990e3bedc90ad66cd

**Comment by Therzok** [comment]
Currently broken/leaking, will probably have to force to instance delegate:
AddClientMessageFilter

**Comment by Therzok** [comment]
https://gist.github.com/Therzok/623f629811ab3e5dd4e65a96f301c923

**Comment by Therzok** [comment]
https://gist.github.com/Therzok/3b4b43d3557fb351d26459cc06892761

**Comment by Therzok** [comment]
https://gist.github.com/Therzok/7df2473a327a882009fca9fea7a1c2c8 latest version

**Comment by Therzok** [comment]
Part 2 of gen, changes to validation in generator.

https://gist.github.com/Therzok/d773e26e39efad8eb9219832bced4c0a

**Comment by Therzok** [comment]
DO NOT MERGE YET. This has a problem with:
```
GtkSharp.AboutDialogActivateLinkFuncNative raw_ret = gtk_about_dialog_set_email_hook(GtkSharp.AboutDialogActivateLinkFuncWrapper.NativeDelegate, data, destroy);
			Gtk.AboutDialogActivateLinkFunc ret = GtkSharp.AboutDialogActivateLinkFuncWrapper.GetManagedDelegate (raw_ret);
			return ret;
```

For example. Target is now no longer set, so GetManagedDelegate will fail.

**Comment by Therzok** [comment]
Diff as of 8c54918 https://gist.github.com/Therzok/2b67fc7f63b7a0e3ef410c7a82aaafe2

**Comment by Therzok** [comment]
Diff for b8d274b https://gist.github.com/Therzok/3372a231bdb96659102a18a88c619114

**Comment by Therzok** [comment]
Diff for ea5f08b https://gist.github.com/Therzok/191c99144158178bb6af8d9b7a1cebf5

**Comment by Therzok** [comment]
Diff for 831efbe https://gist.github.com/Therzok/a67f169546f3db59285b687976d3ddc1

**Comment by Therzok** [comment]
The PR should be ready now.

**Comment by Therzok** [comment]
Final diff:
https://gist.github.com/Therzok/b158ce4dd70ed3bd35120812dd80de3b
 142 files changed, 670 insertions(+), 5743 deletions(-)

**Comment by Therzok** [comment]
Last version:
https://gist.github.com/Therzok/92cf083b3e1157fa9f2a197e189648e7
 144 files changed, 666 insertions(+), 5462 deletions(-)

File: gdk/Display.custom [hunk 1]
```diff
@@ -75,12 +75,8 @@
...
 		public void AddClientMessageFilter (Gdk.Atom message_type, Gdk.FilterFunc func) 
 		{
-			GdkSharp.FilterFuncWrapper func_wrapper = new GdkSharp.FilterFuncWrapper (func);
-			if (!PersistentData.Contains ("client_message_filter_func_list"))
-				PersistentData ["client_message_filter_func_list"] = new ArrayList ();
-			ArrayList func_list = PersistentData ["client_message_filter_func_list"] as ArrayList;
-			func_list.Add (func_wrapper);
-			gdk_display_add_client_message_filter (Handle, message_type == null ? IntPtr.Zero : message_type.Handle, func_wrapper.NativeDelegate, IntPtr.Zero);
+			var gch = GCHandle.Alloc (func);
+			gdk_display_add_client_message_filter (Handle, message_type == null ? IntPtr.Zero : message_type.Handle, GdkSharp.FilterFuncWrapper.NativeDelegate, (IntPtr)gch);
```
Comment (alanmcgovern) [inline gdk/Display.custom]: Can we add a comment that this leaks forever and that there is no `gdk_display_remove_client_message_filter` to allow cleanup to happen?

File: gdk/Window.custom [hunk 2]
```diff
@@ -120,38 +120,42 @@
...
+		if (ret == null)
 			return;
+		GCHandle gch = (GCHandle)ret;
 		hash.Remove (function);
-		gdk_window_remove_filter (Handle, wrapper.NativeDelegate, IntPtr.Zero);
+		gdk_window_remove_filter (Handle, GdkSharp.FilterFuncWrapper.NativeDelegate, (IntPtr)gch);
+		gch.Free();
 	}
 
 #if MANLY_ENOUGH_TO_INCLUDE
```
Comment (alanmcgovern) [inline gdk/Window.custom]: 🤣

File: glade/XML.custom [hunk 3]
```diff
@@ -25,15 +25,20 @@
...
 		public static Glade.XMLCustomWidgetHandler CustomHandler {
 			set {
-				callback_wrapper = new GladeSharp.XMLCustomWidgetHandlerWrapper (value);
-				glade_set_custom_handler(callback_wrapper.NativeDelegate, IntPtr.Zero);
+				if (callback_wrapper != null) {
+					gch.Free ();
+				}
+
+				gch = GCHandle.Alloc (value);
+				glade_set_custom_handler(GladeSharp.XMLCustomWidgetHandlerWrapper.NativeDelegate, (IntPtr)gch);
```
Comment (alanmcgovern) [inline glade/XML.custom]: Don't we need to set `callback_wrapper`? It'll always be null here, right? Should this be changed to:
```
- static Glade.XMLCustomWidgetHandler callback_wrapper = null;
...
...
if (gch.IsAllocated)
    gch.Free ();
```

**Comment by Therzok** [comment]
https://gist.github.com/Therzok/7c79efe0f070277f35c19ef44b9cd770

Latest diff
 144 files changed, 1001 insertions(+), 5650 deletions(-)

File: gdk/Display.custom (hunk 1 above)
Comment (Therzok) [reply gdk/Display.custom]: Done.

File: gdk/Window.custom (hunk 2 above)
Comment (Therzok) [reply gdk/Window.custom]: Yeah, noticed it too

File: glade/XML.custom (hunk 3 above)
Comment (Therzok) [reply glade/XML.custom]: gch.IsAllocated only works after free, afaik

Comment (Therzok) [reply glade/XML.custom]: Fixed.

File: gtk/TreeModelFilter.custom [hunk 4]
```diff
@@ -68,7 +68,7 @@
 			IntPtr[] native_types = new IntPtr [types.Length];
 			for (int i = 0; i < types.Length; i++)
 				native_types [i] = types [i].Val;
-			GCHandle gch = GCHandle.Alloc (func);
+			GCHandle gch = func != null ? GCHandle.Alloc (func) : (GCHandle)IntPtr.Zero;
 			gtk_tree_model_filter_set_modify_func (Handle, n_columns, native_types, GtkSharp.TreeModelFilterModifyFuncWrapper.NativeDelegate, (IntPtr) gch, GLib.DestroyHelper.NotifyHandler);
```
Comment (alanmcgovern) [inline gtk/TreeModelFilter.custom]: Do we want to pass a null destroy callback when func is null? It'd be consistent with everything else, but not a big deal

Comment (alanmcgovern) [reply gtk/TreeModelFilter.custom]: Also, s there a particular reason why we want to keep passing `GtkSharp.TreeModelFilterModifyFuncWrapper.NativeDelegate` to native when the `func` is null?

Comment (Therzok) [reply gtk/TreeModelFilter.custom]: DestroyNotify handles null gchandle, so we're ok:
```
static void ReleaseGCHandle (IntPtr data)
		{
			if (data == IntPtr.Zero)
				return;
			GCHandle gch = (GCHandle) data;
			gch.Free ();
		}

```

Comment (Therzok) [reply gtk/TreeModelFilter.custom]: Hmmm, maybe not. I can send in null. This was custom bound, didn't verify integrity of it

**Pull request merged**